src/tokens/tokens.service.ts

```diff
@@ -0,0 +1,361 @@
+import {
+  AsyncResponse,
+  EthConnectAsyncResponse,
+  EthConnectReturn,
+  TokenBalance,
+  TokenBalanceQuery,
+  TokenBurn,
+  TokenBurnEvent,
+  TokenMint,
+  TokenMintEvent,
+  TokenPool,
+  TokenPoolActivate,
+  TokenPoolEvent,
+  TokenTransfer,
+  TokenTransferEvent,
+  TokenType,
+} from './tokens.interfaces';
+import { decodeHex, encodeHex, packSubscriptionName, unpackSubscriptionName } from './tokens.util';
+
+const TOKEN_STANDARD = 'ERC20';
+const BASE_SUBSCRIPTION_NAME = 'base';
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+const tokenCreateEvent = 'TokenCreate';
+const tokenCreateEventSignature = 'TokenCreate(address,string,string,bytes)';
+const transferEvent = 'Transfer';
+const transferEventSignature = 'Transfer(address,address,uint256)';
+
+@Injectable()
+export class TokensService {
+  private readonly logger = new Logger(TokensService.name);
+
+  baseUrl: string;
+  instancePath: string;
+  instanceUrl: string;
+  topic: string;
+  shortPrefix: string;
+  contractABI: string;
+  contractInstanceUrl: string;
+  stream: EventStream;
+  username: string;
+  password: string;
+
+  constructor(
+    private http: HttpService,
+    private eventstream: EventStreamService,
+    private proxy: EventStreamProxyGateway,
+  ) {}
+
+  configure(
+    baseUrl: string,
+    instancePath: string,
+    topic: string,
+    shortPrefix: string,
+    contractABI: string,
+    username: string,
+    password: string,
+  ) {
+    this.baseUrl = baseUrl;
+    this.instancePath = instancePath;
+    this.instanceUrl = `${baseUrl}${instancePath}`;
+    this.topic = topic;
+    this.shortPrefix = shortPrefix;
+    this.contractABI = contractABI;
+    this.contractInstanceUrl = `${this.baseUrl}/abis/${contractABI}`;
+    this.username = username;
+    this.password = password;
+    this.proxy.addListener(
+      new TokenListener(this.http, this.instanceUrl, this.topic, this.username, this.password),
+    );
+  }
+
+  /**
+   * One-time initialization of event stream and base subscription.
+   */
+  async init() {
+    this.stream = await this.eventstream.createOrUpdateStream(this.topic);
+    await this.eventstream.getOrCreateSubscription(
+      this.instanceUrl,
+      this.stream.id,
+      tokenCreateEvent,
+      packSubscriptionName(this.topic, BASE_SUBSCRIPTION_NAME),
+    );
+  }
+
+  private postOptions(operator: string, requestId?: string) {
+    const from = `${this.shortPrefix}-from`;
+    const sync = `${this.shortPrefix}-sync`;
+    const id = `${this.shortPrefix}-id`;
+
+    const requestOptions: AxiosRequestConfig = {
+      params: {
+        [from]: operator,
+        [sync]: 'false',
+        [id]: requestId,
+      },
+      ...basicAuth(this.username, this.password),
+    };
+
+    return requestOptions;
+  }
+
+  async createPool(dto: TokenPool): Promise<AsyncResponse> {
+    if (dto.type !== undefined && dto.type !== TokenType.FUNGIBLE) {
+      throw new BadRequestException();
```

Comment:
Would be nice to provide a message here

Reply:
Not needed due to new whitelisting config